downloader/download.go

```diff
@@ -0,0 +1,564 @@
+// If the remote resource has not been modified since the given time, the server will
+// respond with a 304.
+func (d *Downloader) IfModifiedSince() *Downloader {
+	d.ifModifiedSince = true
+	return d
+}
+
+// WithLastModified sets the downloader to check the "Last-Modified" header with a HEAD request.
+func (d *Downloader) WithLastModified() *Downloader {
+	d.lastModified = true
+	return d
+}
+
+// WithShelfLife sets the duration after which a file is considered stale, if it has no
+// "Last-Modified" header. If unset, the file will be considered stale by default.
+func (d *Downloader) WithShelfLife(shelfLife time.Duration) *Downloader {
+	d.shelfLife = shelfLife
+	return d
+}
+
+// WithMode sets the file mode for the downloaded file. If not set, the file mode
+// will be taken from the destination file, if it exists before the download.
+func (d *Downloader) WithMode(mode os.FileMode) *Downloader {
+	d.mode = mode
+	return d
+}
+
+// WithHTTPClient sets the http client for the downloader.
+func (d *Downloader) WithHTTPClient(client http.Client) *Downloader {
+	d.httpClient = client
+	return d
+}
+
+// LimitDownloadSize sets the maximum size of the downloaded file,
+// by checking the Content-Length header and monitoring the size again
+// while uncompressing the payload.
+func (d *Downloader) LimitDownloadSize(size int64) *Downloader {
+	d.maxSize = size
+	return d
+}
+
+// getDestInfo returns the modification time and file mode of the destination file.
+func (d *Downloader) getDestInfo() (time.Time, fs.FileMode) {
+	dstInfo, err := os.Stat(d.destPath)
+	switch {
+	case os.IsNotExist(err):
+		return time.Time{}, 0
+	case err != nil:
+		d.logger.Error("Failed to stat destination file", "path", d.destPath, "error", err)
+		return time.Time{}, 0
+	}
+
+	return dstInfo.ModTime(), dstInfo.Mode().Perm()
+}
+
+// checkLastModified returns true if the file seems up to date according to its modification time.
+func (d *Downloader) checkLastModified(ctx context.Context, modTime time.Time) (bool, error) {
+	if !d.lastModified {
+		return false, nil
+	}
+
+	localIsOld := true
+
+	if d.shelfLife != 0 {
+		localIsOld = modTime.Add(d.shelfLife).Before(time.Now())
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodHead, d.url, nil)
+	if err != nil {
+		return false, fmt.Errorf("failed to create HEAD request for %s: %w", d.url, err)
+	}
+
+	resp, err := d.httpClient.Do(req)
+	if err != nil {
+		return false, fmt.Errorf("failed to make HEAD request for %s: %w", d.url, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return false, BadHTTPCodeError{d.url, resp.StatusCode}
+	}
+
+	remoteLastModified := resp.Header.Get("Last-Modified")
+	if remoteLastModified == "" {
+		if !localIsOld {
+			d.logger.Debug("No last modified header, but local file is not old",
+				"url", d.url, "path", d.destPath, "shelflife", d.shelfLife)
+			return true, nil
+		}
+	}
+
+	if remoteLastModified == "" {
```

Comment:
The condition is same as the precedent. Is this intended ?

code seems to be functional though, so maybe it is.

Reply:
thanks i'll attach the two if